File: session/02_double_dispatch/haskell/02-paper_scissor_rock.hs
-- We’ll use ADT Element Rock  Paper Scissor
-- to represent the three choices.
data Element = Paper | Scissor | Rock 
  deriving (Show)

beats :: Element -> Element -> String
beats Paper   Scissor = "paper loses"
beats Paper   Rock    = "paper wins"
beats Scissor Paper   = "scissor wins"
beats Scissor Rock    = "scissor loses"
beats Rock    Paper   = "rock loses"
beats Rock    Scissor = "rock wins"
beats _       _       = "draw"

main :: IO ()
main = do
  print $ "paper beats paper = "    ++ (Paper   `beats` Paper)
  print $ "paper beats scissor = "  ++ (Paper   `beats` Scissor)
  print $ "paper beats rock = "     ++ (Paper   `beats` Rock)
  print $ "scissor beats paper = "  ++ (Scissor `beats` Paper)
  print $ "scissor beats scissor = "++ (Scissor `beats` Scissor)
  print $ "scissor beats rock = "   ++ (Scissor `beats` Rock)
  print $ "rock beats paper = "     ++ (Rock    `beats` Paper)
  print $ "rock beats scissor = "   ++ (Rock    `beats` Scissor)
  print $ "rock beats rock = "      ++ (Rock    `beats` Rock)
 

File: session/02_double_dispatch/02-paper-scissor-rock.scala
trait Beatable {
	def beats(other: Beatable): String
}

case object Paper extends Beatable {
	override def beats(other: Beatable): String = other match {
	  case Rock => "paper wins"
    case Scissor => "paper loses"
    case Paper => "draw"
	}
}

case object Rock extends Beatable {
	override def beats(other: Beatable): String = other match {
	  case Rock => "draw"
    case Scissor => "rock wins"
    case Paper => "rock loses"
	}
}
case object Scissor extends Beatable {
	override def beats(other: Beatable): String = other match {
	  case Rock => "scissor loses"
    case Scissor => "draw"
    case Paper => "scissor wins"
	}
}

val paper = Paper
val rock = Rock
val scissor = Scissor

println("paper.beats(paper) = " + paper.beats(paper))
println("paper.beats(scissor) = " + paper.beats(scissor))
println("paper.beats(rock) = " + paper.beats(rock))
println("scissor.beats(paper) = " + scissor.beats(paper))
println("scissor.beats(scissor) = " + scissor.beats(scissor))
println("scissor.beats(rock) = " + scissor.beats(rock))
println("rock.beats(paper) = " + rock.beats(paper))
println("rock.beats(scissor) = " + rock.beats(scissor))
println("rock.beats(rock) = " + rock.beats(rock))
